bpf/libbpf/libbpf.go

```diff
@@ -0,0 +1,200 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package libbpf
+
+import (
+	"fmt"
+	"unsafe"
+
+	"github.com/projectcalico/felix/bpf"
+)
+
+// #include "libbpf_api.h"
+import "C"
+
+type Obj struct {
+	obj *C.struct_bpf_object
+}
+
+type TCOpts struct {
+	opts C.struct_bpf_tc_opts
+}
+
+type Map struct {
+	bpfMap *C.struct_bpf_map
+	bpfObj *C.struct_bpf_object
+}
+
+const MapTypeProgrArray = C.BPF_MAP_TYPE_PROG_ARRAY
+
+type QdiskHook string
+
+const (
+	QdiskIngress QdiskHook = "ingress"
+	QdiskEgress  QdiskHook = "egress"
+)
+
+func (m *Map) Name() string {
+	name, err := C.bpf_map__name(m.bpfMap)
+	if err != nil {
+		return ""
+	}
+	return C.GoString(name)
+}
+
+func (m *Map) Type() int {
+	mapType, err := C.bpf_map__type(m.bpfMap)
+	if err != nil {
+		return -1
+	}
+	return int(mapType)
+}
+
+func (m *Map) SetPinPath(path string) error {
+	cPath := C.CString(path)
+	defer C.free(unsafe.Pointer(cPath))
+	err := C.bpf_map__set_pin_path(m.bpfMap, cPath)
+	if err != 0 {
+		return fmt.Errorf("pinning map failed %w", err)
+	}
+	return nil
+}
+
+func OpenObject(filename string) (*Obj, error) {
+	bpf.IncreaseLockedMemoryQuota()
+	cFilename := C.CString(filename)
+	defer C.free(unsafe.Pointer(cFilename))
+	obj, err := C.bpf_obj_open(cFilename)
+	if obj == nil || err != nil {
+		return nil, fmt.Errorf("error opening libbpf object %w", err)
+	}
+	return &Obj{obj: obj}, nil
+}
+
+func (o *Obj) Load() error {
+	_, err := C.bpf_obj_load(o.obj)
+	if err != nil {
+		return fmt.Errorf("error loading object %w", err)
+	}
+	return nil
+}
+
+func (o *Obj) FirstMap() (*Map, error) {
+	bpfMap, err := C.bpf_map__next(nil, o.obj)
+	if bpfMap == nil || err != nil {
+		return nil, fmt.Errorf("error getting first map %w", err)
+	}
+	return &Map{bpfMap: bpfMap, bpfObj: o.obj}, nil
+}
+
+func (m *Map) NextMap() (*Map, error) {
+	bpfMap, err := C.bpf_map__next(m.bpfMap, m.bpfObj)
+	if err != nil {
+		return nil, fmt.Errorf("error getting next map %w", err)
+	}
+	if bpfMap == nil {
+		return nil, nil
+	}
+	return &Map{bpfMap: bpfMap, bpfObj: m.bpfObj}, nil
+}
+
+func (o *Obj) AttachClassifier(secName, ifName, hook string) (*TCOpts, error) {
+	isIngress := 0
+	cSecName := C.CString(secName)
+	cIfName := C.CString(ifName)
+	defer C.free(unsafe.Pointer(cSecName))
+	defer C.free(unsafe.Pointer(cIfName))
+	ifIndex, err := C.if_nametoindex(cIfName)
+	if err != nil {
+		return nil, err
+	}
+
+	if hook == string(QdiskIngress) {
+		isIngress = 1
+	}
+
+	opts, err := C.bpf_tc_program_attach(o.obj, cSecName, C.int(ifIndex), C.int(isIngress))
+	if err != nil {
+		return nil, fmt.Errorf("Error attaching tc program %w", err)
+	}
+	return &TCOpts{opts: opts}, nil
```

Comment:
You could do the prog ID lookup here and return that as an `int`; would that be a nicer API?